draft-ietf-taps-interface.md

```diff
@@ -503,44 +503,49 @@ implemented over UDP.
 This property specifies whether the application needs to use a transport
 protocol that ensures that all data is received on the other side without
 corruption. This also entails being notified when a Connection is closed or
-aborted. The recommended default is to enable Reliable Data Transfer.
+aborted. The recommended default is to Require Reliable Data Transfer.
+
+### Preservation of Message Boundaries {#prop-boundaries}
+
+This property specifies whether the application needs to use a transport
+protocol that preserves message boundaries. The recommended default
+is to Prefer Preservation of Message Boundaries.
 
 ### Configure per-Message reliability {#prop-partially-reliable}
 
 This property specifies whether an application considers it useful to indicate
 its reliability requirements on a per-Message basis. This property applies to
-Connections and Connection Groups. The recommended default is to not have this
+Connections and Connection Groups. The recommended default is to Prefer this
 option.
 
 ### Preservation of data ordering {#prop-ordering}
 
 This property specifies whether the application wishes to use a transport
 protocol that can ensure that data is received by the application on the other
-end in the same order as it was sent. The recommended default is to preserve
-data ordering.
+end in the same order as it was sent. The recommended default is to Require
+Preservation of data ordering.
 
 ### Use 0-RTT session establishment with an idempotent Message {#prop-0rtt}
 
 This property specifies whether an application would like to supply a Message to
 the transport protocol before Connection establishment, which will then be
 reliably transferred to the other side before or during Connection
 establishment, potentially multiple times. See also {{msg-idempotent}}. The
-recommended default is to not have this option.
+recommended default is to Prefer this option.
 
 ### Multistream Connections in Group {#prop-multistream}
 
 This property specifies that the application would prefer multiple Connections
 within a Connection Group to be provided by streams of a single underlying
-transport connection where possible. The recommended default is to not have
+transport connection where possible. The recommended default is to Prefer have
```

Comment:
Same comment as above.

(It actually seems to me that smarter TAPS implementations could even extract the preferences/requirements of an application from static code analysis. An app that never clones connections doesn't need multistreaming, as an app that never sets deadlines/reliability doesn't need partial reliability)

Reply:
The above (supporting per-message reliability) is a matter of "does the application care", as in "will it use this functionality"?  However this is a very different case!  This is about being allowed to group them without the application caring about it - I think we definitely want to do this, to benefit from multistreaming whenever possible. The benefits can be large even without app involvement, as both Connections get a single congestion control context.

For our API design, let's not presume that the underlying system can see into the future by analyzing the app code. Sequentially, this is before an application could do "clone".

Reply:
It will be difficult to do multi-streaming without explicit action by the application (calling clone).

The "automagic multi-streaming mode" works for my connection pool proposal (I should just write an PR for that), but not for regular connections. We tried searching for "matching connections" on ```socketconnect()```in our Socket Intents Prototype, but this was really tedious and error prone.
So requiring/preferring it really means "the application may want to call ```clone()```"

Reply:
"Automagic" streaming mode is possible and has always been part of the model - see minset. Also, NEAT is able to do this. For details, see:
F. Weinrank and M. Tüxen. “Transparent Flow Mapping for NEAT“. In Workshop on Future of Internet Transport (FIT 2017), Stockholm, June 2017.
http://dl.ifip.org/db/conf/networking/networking2017/1570349380.pdf

Reply:
@mwelzl in case of NEAT, this is explicitly negotiated with the peer, which is a contradiction to the TAPS requirement to work with a non-TAPS peer.
If we accept that Transparent Flow Mapping is a protocol feature that might be present, I am unsure how does this will interact with an application that explicitly uses ```clone()``` to initiate another stream? 
 - Do we need a parameter to turn off Transparent Flow Mapping to allow manual flow mapping with clone?
 - What about protocols like HTTP/3 that have a fixed stream-id to protocol semantic mapping?

Reply:
The negotiation happens in SCTP - NEAT itself doesn't know or care about the negotiation, it just locally tries to use SCTP with all necessary features. So saying this contradicts TAPS is like saying "using MPTCP contradicts TAPS because it requires negotiation"; it's just a local thing to try. But the idea is to definitely allow (even encourage!) people implementing TAPS systems to support this, as it can yield a lot of benefits (reduced FCTs, less overall delay from less congestion control interactions, ..) without the application having to worry about everything.

About your questions:
1) No, you can still do this. (Why do you think you couldn't?)

2) Does it, *still*? Sigh... well I guess it's just a matter of giving a stream number to a Preconnection. That should be a protocol-specific property, I guess. Moreover, the requirement level for "Multistream Connections in Groups" will have to be "Require" if you want that level of control.

Reply:
HTTP/3 still does, but the mapping has become less ugly. I think we should not worry too much here as HTTP ugliness can be easily hid with connection pools and pseudo transports. An average application programmer has no change of implementing HTTP/1.1 + 2 + 3 including negotiation and fallback anyway.